code/modules/research/xenobiology/crossbreeding/warping.dm

```diff
@@ -0,0 +1,1227 @@
+	first_person = null
+	second_person = null
+	return ..()
+
+
+/*Acts as a bunker making anything into the rune immune to outside explosions and gas leaks. doesn't work if the epicenter of the explosion is IN the rune */
+
+
+/obj/item/slimecross/warping/oil
+	colour = "oil"
+	runepath = /obj/effect/warped_rune/oilspace
+	effect_desc = "protects anything on the rune from explosions outside the rune.."
+
+
+/obj/effect/warped_rune/oilspace
+	icon_state = "oil_rune"
+	desc = "The world is ending, but we have one last trick up our sleeve, we will survive."
+	///used to remember the oilspace_bunker specific to this rune
+	var/list/bunker_list
+
+
+/obj/effect/oilspace_bunker //we'll surround the rune with these so it "blocks" nearby explosions. Although only the rune itself is 100% protected
+	mouse_opacity = MOUSE_OPACITY_TRANSPARENT
+	invisibility = TRUE //no one can see those
+	anchored = TRUE
+	explosion_block = INFINITY
+	CanAtmosPass = ATMOS_PASS_NO //will also try to stop gas from exiting the rune to keep the pressure and air of the tile.Doesn't always work
+	move_resist = INFINITY //we need it to stay in the same place until it's deleted.
+	resistance_flags = INDESTRUCTIBLE | LAVA_PROOF | FIRE_PROOF | UNACIDABLE | ACID_PROOF
+
+
+/obj/effect/warped_rune/oilspace/Initialize()
+	. = ..()
+	for(var/turf/bunker_turf in range(1,rune_turf))
+		var/obj/effect/oilspace_bunker/bunker_wall = new /obj/effect/oilspace_bunker(bunker_turf)
+		LAZYADD(bunker_list, bunker_wall)
+
+
+/obj/effect/warped_rune/oilspace/Destroy()
+	for(var/obj/effect/oilspace_bunker/bunker_wall in range(1,rune_turf))
+		if(bunker_wall in bunker_list)
+			LAZYREMOVE(bunker_list, bunker_wall)
+			qdel(bunker_wall)
+			return ..()
+
+
+/* Used to teleport anything over it to a unique room similar to hilbert's hotel.*/
+
+
+/obj/item/slimecross/warping/rainbow
+	colour = "rainbow"
+	effect_desc = "draws a rune that will teleport anything above it "
+	runepath = /obj/effect/warped_rune/rainbowspace
+	drawing_time = 150
+	storing_time = 100
+
+/obj/effect/warped_rune/rainbowspace
+	icon_state = "rainbow_rune"
+	desc = "This is where I go when I want to be alone. Yet they keep clawing at the walls until everything crumbles."
+	///current x,y,z location of the reserved space for the rune room
+	var/datum/turf_reservation/room_reservation
+	///the template of the warped_room map
+	var/datum/map_template/warped_room/rune_room
+	///list of people that teleported into the rune_room. The room will dissapear if the list is empty and the rune is destroyed.
+	var/list/customer_list
+
+
+/obj/effect/warped_room_exit
+	name = "warped_rune"
+	icon = 'icons/obj/slimecrossing.dmi'
+	icon_state = "rainbow_rune"
+	desc = "Use this rune if you want to leave this place. You will have to leave eventually."
+	move_resist = INFINITY
+	anchored = TRUE
+	///where the rune will teleport you back.
+	var/turf/exit_turf
+	///rune linked to the exit rune
+	var/obj/effect/warped_rune/rainbowspace/enter_rune
```

Comment:
Clean these variables on `Destroy()` and register on their destruction in case they may be destroyed before the effect is.